packages/pigeon/lib/src/dart/dart_generator.dart

```diff
@@ -288,6 +303,49 @@ class DartGenerator extends StructuredGenerator<DartOptions> {
     });
   }
 
+  @override
+  void writeClassEquality(
+    DartOptions generatorOptions,
+    Root root,
+    Indent indent,
+    Class classDefinition, {
+    required String dartPackageName,
+  }) {
+    indent.writeln('@override');
+    indent.writeln('// ignore: avoid_equals_and_hash_code_on_mutable_classes');
+    indent.writeScoped('bool operator ==(Object other) {', '}', () {
+      indent.writeScoped(
+          'if (other is! ${classDefinition.name} || other.runtimeType != runtimeType) {',
+          '}', () {
+        indent.writeln('return false;');
+      });
+      indent.writeScoped('if (identical(this, other)) {', '}', () {
+        indent.writeln('return true;');
+      });
+      indent.writeScoped('return ', '', () {
+        indent.format(
+            classDefinition.fields
+                .map((NamedType field) => field.type.baseName == 'List' ||
+                        field.type.baseName == 'Float64List' ||
+                        field.type.baseName == 'Int32List' ||
+                        field.type.baseName == 'Int64List' ||
+                        field.type.baseName == 'Uint8List'
+                    ? '_listEquals(${field.name}, other.${field.name})'
+                    : field.type.baseName == 'Map'
+                        ? '_mapEquals(${field.name}, other.${field.name})'
+                        : '${field.name} == other.${field.name}')
```

Comment:
You would also then have the option of just unconditionally using `_deepEquals` here. There is the argument that doing the type check here at generation time makes the runtime check slightly more efficient, but absent an actual use case where we someone cares about micro-optimizing equality checks (our own use cases are for tests, where it definitely doesn't matter) I would lean toward making the generator code simpler, but either way is fine with me.

Reply:
Makes sense to me, I just don't like losing the type safety. It will obviously be added in the method itself though, so it's not a big deal.